### PR TITLE
Factor out code for windows abi tags.

### PR DESCRIPTION
**Pull request by bredelings**

### CODE REVIEW

File: mesonbuild/dependencies/misc.py [hunk 1]
```diff
@@ -301,13 +301,51 @@ def modname_from_filename(self, filename):
             modname = modname[3:]
         return modname
 
+    def compiler_tag(self):
+        tag = None
+        if mesonlib.for_windows(self.want_cross, self.env):
+            comp_ts_version = self.env.detect_cpp_compiler(self.want_cross).get_toolset_version()
+            compiler_ts = comp_ts_version.split('.')
+            #FIXME - what about other compilers?
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/misc.py]: __[Flake8]__

[E265] block comment should start with '# '

File: mesonbuild/dependencies/misc.py [hunk 2]
```diff
@@ -301,13 +301,51 @@ def modname_from_filename(self, filename):
...
+            tag = 'vc{}{}'.format(compiler_ts[0], compiler_ts[1])
+
+            tag = '-' + tag;
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/misc.py]: __[Flake8]__

[E703] statement ends with a semicolon

File: mesonbuild/dependencies/misc.py [hunk 3]
```diff
@@ -301,13 +301,51 @@ def modname_from_filename(self, filename):
...
+            # - Cygwin:   leaves off -mt but libraries are multithreading-aware.
+            return ''
+
+    def version_tag(self):
+        return self.version.replace('.', '_')
+
+    # FIXME - how to handle different distributions, e.g. for Mac? Currently we handle homebrew and macports, but not fink.
+    def abi_tag(self):
+        if mesonlib.for_windows(self.want_cross, self.env):
+            tag = self.compiler_tag()+self.threading_tag()
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/misc.py]: __[Flake8]__

[E226] missing whitespace around arithmetic operator

**Comment by bredelings** [comment]
I will look into it soon. I'm relying on the appveyor builds for feedback
on windows, and they take a while.

File: mesonbuild/dependencies/misc.py [hunk 4]
```diff
@@ -301,20 +305,67 @@ def modname_from_filename(self, filename):
...
+    def version_tag(self):
+        return '-' + self.version.replace('.', '_')
+
+    def debug_tag(self):
+        return '-gd' if self.is_debug else ''
+
+    # FIXME - how to handle different distributions, e.g. for Mac? Currently we handle homebrew and macports, but not fink.
+    def abi_tag(self):
+        if mesonlib.for_windows(self.want_cross, self.env):
+            # PROBLEM: mingw just uses self.threading_tag()
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/misc.py]: __[Flake8]__

[W291] trailing whitespace

**Comment by bredelings** [comment]
Fixes #2990

**Comment by bredelings** [comment]
Fixes #2631

**Comment by bredelings** [comment]
@nirbheek Any comments?

**Comment by nirbheek** [comment]
I can't review in detail, but this sounds good to me in principle. Maybe @NickeZ can review?

**Pull request merged**